### ISSUE TITLE
Grain Timer stop firing after a while

### ISSUE THREAD

**Issue by sc-starman** [dotnet/orleans#8800]
Using Orleans version 7.2.3 for a while for a group of long-lived grain with high throughput calls and now I noticed the grain timer which I registered at the activation will stop firing after 5-6 hours without any notice or log.

Registration code is like below and I store _publishOrderbookTimer  in grain class:
```
        public override async Task OnActivateAsync(CancellationToken cancellationToken)
        {
            _logger.LogInformation("Activation of Matching Engine Happend, Pair={Pair}", Pair);
            _publishOrderbookTimer = RegisterTimer(PublishOrderbook, null, TimeSpan.FromSeconds(1), TimeSpan.FromMilliseconds(100));
            await base.OnActivateAsync(cancellationToken);
        }

        public override Task OnDeactivateAsync(DeactivationReason reason, CancellationToken cancellationToken)
        {
            _logger.LogInformation("Deactivation of Matching Engine Happend, Pair={Pair}, Reason={reason}", Pair, reason);
            if (_publishOrderbookTimer is not null)
            {
                _publishOrderbookTimer.Dispose();
                _publishOrderbookTimer = null;
            }
            return base.OnDeactivateAsync(reason, cancellationToken);
        }
```

I can confirm that the grain is still alive and functioning correctly but the timer is not.

**Comment by sc-starman** [comment]
Upgraded to 8.0.0-rc2 but still got the same problem.
Still, there is no log or evidence that I could use to find out what happened.

**Comment by ReubenBond** [comment]
How are you verifying that the timer is not firing? Could you show the relevant code? When you say that the grain continues to function, I assume this means that there are no logs showing that the grain has been deactivated. Is that correct?

**Comment by sc-starman** [comment]
This is the timer function code:
```
        private async Task PublishOrderbook(object arg)
        {
            if (_orderbookUpdatedTimestamp <= _orderbookSentTimestamp)
            {
                if ((DateTime.UtcNow - _orderbookUpdatedTimestamp).TotalSeconds == 5)
                    _logger.LogCritical("Orderbook not Updated Recently, Pair={Pair}, LastUpdateInSeconds={LastUpdateInSeconds}, Hostname={Hostname}",
                        Pair, (DateTime.UtcNow - _orderbookUpdatedTimestamp).TotalSeconds, Dns.GetHostName());
                return;
            }

            _orderbookSentTimestamp = DateTime.UtcNow;

                    _logger.LogCritical("Orderbook Updated Successfuly, Pair={Pair}, LastUpdateInSeconds={LastUpdateInSeconds}, Hostname={Hostname}",
                        Pair, (DateTime.UtcNow - _orderbookUpdatedTimestamp).TotalSeconds, Dns.GetHostName());


            int count = 40;
            var book = _engine.Book;
            var asks = book.AskSide;
            var bids = book.BidSide;

            _orderBook = new OrderBookStream()
            {
                Pair = Pair,
                Asks = asks.Take(count).ToDictionary(a => (decimal)a.Value.Price, a => (decimal)a.Value.Quantity),
                Bids = bids.Take(count).ToDictionary(a => (decimal)a.Value.Price, a => (decimal)a.Value.Quantity),
                BestAskPrice = book.BestAskPrice,
                BestAskQuantity = book.BestAskQuantity,
                BestBidPrice = book.BestBidPrice,
                BestBidQuantity = book.BestBidQuantity,
                BestStopAskPrice = book.BestStopAskPrice,
                BestStopBidPrice = book.BestStopBidPrice,
                Timestamp = book.LastBookUpdate
            };

            var subject = $"{typeof(OrderBookStream).FullName}_{Pair}";


            await _natsCommand.PublishAsync(subject, _orderBook);
        }
```
Yeah not only I'm sure the grain is not deactivated (no log) but also I have a grain method that I can query the difference between when the timer fired and how much time passed from the grain activation. Everything seems fine except the timer.

**Comment by sc-starman** [comment]
Further investigation, I realized that this occurred on the silo level not at the grain level, which means that all timers which belong to each grain on a specific silo won't fire at all. Even if I registered them again (when the grain was already active), they would fire only for the first time.

**Comment by ReubenBond** [comment]
I believe this was fixed by #8864

Please open a new issue pointing to this one if the issue persists after upgrading to v8.1.0
